### ISSUE TITLE
Possibly false positive "non-void function does not return a value in all control paths" warning 

### ISSUE THREAD

**Issue by sergey-senozhatsky** [llvm/llvm-project#59418]
The following sample code
```c++
long offset(long v)
{
	if (!v)
		return 0;
	if (v > 0)
		return 1;
	if (v < 0)
		return -1;
}
```
produces
```console
warning: non-void function does not return a value in all control paths [-Wreturn-type]
```

So does 
```c++
long offset(long v)
{
	if (v == 0)
		return 0;
	if (v > 0)
		return 1;
	if (v < 0)
		return -1;
}
```

Ditto with `-O2`, `-Os`.

```console
$ clang --version
Debian clang version 14.0.6-2
Target: x86_64-pc-linux-gnu
Thread model: posix
```

**Comment by ribalda** [comment]
The warning is also at trunk https://godbolt.org/z/cdsq1qava

**Comment by DimitryAndric** [comment]
FWIW, gcc produces the same warning :)

```text
% gcc-12 -c pr59418-1.cpp
pr59418-1.cpp: In function 'long int offset(long int)':
pr59418-1.cpp:9:1: warning: control reaches end of non-void function [-Wreturn-type]
    9 | }
      | ^
```

**Comment by mydeveloperday** [comment]
msvc and icc  give similar warnings so clang is in good company.

```
Output of x86-64 icc 2021.7.1 (Compiler #1)
icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
<source>(9): warning #1011: missing return statement at end of non-void function "offset"
  }
```

``
example.c
z:\tmp\compiler-explorer-compiler2022119-8890-iz8e1d.l2ekl\example.c(9) : warning C4715: 'offset': not all control paths return a value
Compiler returned: 0
```

**Comment by shafik** [comment]
This is straight forward to work around. Rearrange the code so that you have early exits and a default return value e.g.:

```cpp
long offset(long v) {    
    if (v > 0)
        return 1;
    if (v < 0)
        return -1;
 
   return 0;
}
```

This also feel a bit easier to reason about and is more obviously safe.

The diagnostic could still probably be improved but considering we can work around it so simply I think not high priority.

**Comment by tbaederr** [comment]
> The diagnostic could still probably be improved

We could add a fixit for `__builtin_unreachable()` or `std::unreachable()` in C++23 (and recognize to not emit the diagnostic).

Also that diagnostic should totally be a hard error but meh.

**Comment by AaronBallman** [comment]
Duplicate of #111509
